fix: reduce float and half-float values to their stored precision

salvatore-campagna opened the pull request:
This is a fix for #82867 and #81529.

The bug is a result of reducing precision of `float` values. If the filed the range aggregation is applied to is a `long` field we try to reduce the precision of the rage `from` and/or `to` values to `long`. This results in an exception at parsing time happening when parsing a `long` with decimal positions.

Here we parse `FLOAT` and `HALF-FLOAT` values actually reducing their precision. For other types, including `long` we do nothing, or better, we do not do any "precision-reduction" action in method `reduceToStoredPrecision`.

salvatore-campagna committed:
fix: reduce float and half-float values to their stored precision

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
Update docs/changelog/83213.yaml

salvatore-campagna committed:
fix: restore skip version changed by mistake

nik9000 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/40_range.yml:
```diff
@@ -199,6 +199,45 @@ setup:
+---
+"Double range on long field":
+  - skip:
+      version: " - 7.16.99"
```
I think this has to be 8.0.99 or the BWC tests will fail while you backport.

salvatore-campagna committed:
fix: just skip versions before 8.1.0

salvatore-campagna committed:
test: include a test for min and max long range bounds

This PR, was intended to fix #82867 but, as a side effect,
it also fixes #81529. These unit and yaml test cover the
use case reposrted by #81529.

salvatore-campagna committed:
chore: restore original indentation

salvatore-campagna committed:
fix: restore test removed by mistake

salvatore-campagna committed:
fix: use a more meaningful test name

salvatore-campagna committed:
fix: code format violations

salvatore-campagna committed:
fix: there is just one document in the bucket

salvatore-campagna committed:
Merge branch 'master' into fix/range-agg-with-decimal

salvatore-campagna commented:
@elasticmachine run elasticsearch-ci/docs

not-napoleon reviewed (approved):
LGTM.  Thanks for jumping on this.

salvatore-campagna commented:
The bug is a result of reducing precision of float values. If the filed the range aggregation is applied to is a long field, we try to reduce the precision of the rage `from` and/or `to` values to `long`. This results in an exception at parsing time happening when parsing a long with decimal positions.

Here we parse FLOAT and HALF-FLOAT values actually reducing their precision. For other types, including `long` we do nothing, or better, we do not do any "precision-reduction" action in method reduceToStoredPrecision.

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.0  | Commit could not be cherrypicked due to conflicts |
| ❌ |  7.17  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 83213`

salvatore-campagna commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83410">](https://github.com/elastic/elasticsearch/pull/83410)|
|✅|7.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83411">](https://github.com/elastic/elasticsearch/pull/83411)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

salvatore-campagna commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83415">](https://github.com/elastic/elasticsearch/pull/83415)|
|✅|7.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83416">](https://github.com/elastic/elasticsearch/pull/83416)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

salvatore-campagna commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/83415">](https://github.com/elastic/elasticsearch/pull/83415)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

JWilsonElastic commented:
Just swinging by to ask if this backport ultimately failed. It looks like the backport PR failed, but wanted to confirm there was no successful backport for 7.x.
